fred2/musicplayerdlg.cpp

```diff
@@ -0,0 +1,214 @@
+// MusicPlayerDlg.cpp : implementation file
+//
+
+#include "stdafx.h"
+#include "FRED.h"
+#include "MainFrm.h"
+#include "musicplayerdlg.h"
+#include "TextViewDlg.h"
+#include "mod_table/mod_table.h"
+#include "sound/audiostr.h"
+
+#ifdef _DEBUG
+#undef THIS_FILE
+static char THIS_FILE[] = __FILE__;
+#endif
+
+music_player_dlg::music_player_dlg(CWnd* pParent /*=nullptr*/)
+	: CDialog(music_player_dlg::IDD, pParent)
+{
+
+	m_music_item = "";
+	m_music_id = -1;
+	m_cursor_pos = -1;
+	m_autoplay = FALSE;
+	m_num_music_files = 0;
+
+}
+
+void music_player_dlg::DoDataExchange(CDataExchange* pDX)
+{
+	CDialog::DoDataExchange(pDX);
+
+	DDX_Control(pDX, IDC_MUSIC_LIST, m_music_list);
+	DDX_Check(pDX, IDC_MUSIC_AUTOPLAY, m_autoplay);
+
+}
+
+BEGIN_MESSAGE_MAP(music_player_dlg, CDialog)
+
+ON_LBN_SELCHANGE(IDC_MUSIC_LIST, OnSelMusicList)
+ON_BN_CLICKED(IDC_BUTTON_PLAY_MUSIC, OnPlay)
+ON_BN_CLICKED(IDC_BUTTON_STOP_MUSIC, OnStop)
+ON_BN_CLICKED(IDC_BUTTON_NEXT_MUSIC, OnNextTrack)
+ON_BN_CLICKED(IDC_BUTTON_PREV_MUSIC, OnPreviousTrack)
+ON_BN_CLICKED(IDC_BUTTON_MUSIC_TBL, OnMusicTbl)
+ON_BN_CLICKED(IDC_MUSIC_AUTOPLAY, OnAutoplay)
+ON_WM_CLOSE()
+
+END_MESSAGE_MAP()
+
+BOOL music_player_dlg::Create()
+{
+	BOOL r;
+	r = CDialog::Create(IDD, Fred_main_wnd);
+
+	m_play_bm.LoadBitmap(IDB_PLAY);
+	((CButton*)GetDlgItem(IDC_BUTTON_PLAY_MUSIC))->SetBitmap(m_play_bm);
+
+	m_stop_bm.LoadBitmap(IDB_STOP);
+	((CButton*)GetDlgItem(IDC_BUTTON_STOP_MUSIC))->SetBitmap(m_stop_bm);
+
+	m_next_bm.LoadBitmap(IDB_NEXT);
+	((CButton*)GetDlgItem(IDC_BUTTON_NEXT_MUSIC))->SetBitmap(m_next_bm);
+
+	m_prev_bm.LoadBitmap(IDB_PREV);
+	((CButton*)GetDlgItem(IDC_BUTTON_PREV_MUSIC))->SetBitmap(m_prev_bm);
+
+	return r;
+}
+
+BOOL music_player_dlg::OnInitDialog()
+{
+	CDialog::OnInitDialog();
+
+	m_music_list.ResetContent();
+
+	SCP_vector<SCP_string> files;
+	cf_get_file_list(files, CF_TYPE_MUSIC, "*", CF_SORT_NAME);
+
+	for (auto& file : files) {
+		bool addItem = true;
+
+		// check if the file was asked to be ignored in game_settings
+		for (auto& thisFile : Ignored_music_player_files) {
+			if (!stricmp(thisFile.c_str(), file.c_str())) {
+				addItem = false;
+				break;
+			}
+		}
+
+		//add item if not ignored
+		if (addItem) {
+			m_music_list.AddString(file.c_str());
+			m_num_music_files++;
+		}
+	}
+
+	return TRUE;
+}
+
+void music_player_dlg::PlayMusic()
+{
+	StopMusic();
+	
+	// cfile strips the extension so first let's try .wav
+	SCP_string thisMusic = m_music_item + ".wav";
+	m_music_id = audiostream_open(thisMusic.c_str(), ASF_EVENTMUSIC);
+
+	// if no file was loaded then it must be .ogg
+	if (m_music_id < 0) {
+		thisMusic = m_music_item + ".ogg";
+		m_music_id = audiostream_open(thisMusic.c_str(), ASF_EVENTMUSIC);
+	}
+
+	if (m_music_id >= 0) {
+		audiostream_play(m_music_id, 1.0f, 0);
+	} else {
+		thisMusic = m_music_item;
+		mprintf(("FRED failed to open music file %s in the music player\n", thisMusic.c_str()));
```

Comment:
Is there a way to show this on the dialog window?  I'm not as good at MFC as I am with qt, but it could be good to have visual feedback, since a warning would probably be overkill.

Reply:
Honestly, I'd rather do a warning. The music player window isn't very big and would make a sentence like that hard to read even if I could find a place for it.

Reply:
I think a warning would be good then, since the mission maker will not have to check the debug log to see if the song failed to open.